### PR TITLE
Feature/patro23-226 create budget validator

### PR DESCRIPTION
**Pull request by a-lipert**
https://tracker.intive.com/jira/browse/PATRO23-226

### CODE REVIEW

**Comment by mateuszgiedzinski** [comment]
Ok for now code review can't be performed because of too many changes between dev and this branch. We need to wait for CreateBudgetAggregate changes.

File: src/modules/budget/application/Budget/CreatingBudget/CreateBudget.cs [hunk 1]
```diff
@@ -14,8 +15,9 @@ namespace Intive.Patronage2023.Modules.Budget.Application.Budget.CreatingBudget;
 /// <param name="Period">Budget time span.</param>
 /// <param name="Description">Description.</param>
 /// <param name="IconName">Budget icon identifier.</param>
+/// <param name="QueryBus">Query bus.</param>
 
-public record CreateBudget(Guid Id, string Name, Guid UserId, Money Limit, Period Period, string Description, string IconName) : ICommand;
+public record CreateBudget(Guid Id, string Name, Guid UserId, Money Limit, Period Period, string Description, string IconName, IQueryBus QueryBus) : ICommand;
```
Comment (mateuszmuszynski) [inline src/modules/budget/application/Budget/CreatingBudget/CreateBudget.cs:20]: We shouldn't include query bus in command body.

File: src/modules/budget/infrastructure/Domain/Rules/BudgetNameUniquenessBusinessRule.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	private readonly string name;
+
+	/// <summary>
+	/// Initializes a new instance of the <see cref="BudgetNameUniquenessBusinessRule"/> class.
+	/// </summary>
+	/// <param name="queryBus">Query bus.</param>
+	/// <param name="budgetDbContext">DB Context.</param>
+	/// <param name="executionContextAccessor"> User context accessor.</param>
+	/// <param name="name">Budget name.</param>
+	public BudgetNameUniquenessBusinessRule(IQueryBus queryBus, BudgetDbContext budgetDbContext, IExecutionContextAccessor executionContextAccessor, string name)
```
Comment (mateuszmuszynski) [inline src/modules/budget/infrastructure/Domain/Rules/BudgetNameUniquenessBusinessRule.cs:25]: Query bus is not used.

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

File: src/modules/budget/application/Budget/CreatingBudget/CreateBudget.cs (hunk 1 above)
Comment (a-lipert) [reply src/modules/budget/application/Budget/CreatingBudget/CreateBudget.cs]: Sorry, I misunderstood the requirements.

File: src/modules/budget/application/Budget/CreatingBudget/CreateBudgetValidator.cs [hunk 3]
```diff
@@ -14,12 +16,40 @@ public class CreateBudgetValidator : AbstractValidator<CreateBudget>
...
+		this.RuleFor(budget => budget.Limit.Value)
+		.GreaterThan(0);
+		this.RuleFor(budget => budget.Limit.Currency)
+		.NotEmpty();
+	}
+
+	private async Task<bool> AnyExistingBudget(string name, IExecutionContextAccessor executionContextAccessor)
+	{
+		bool anyExistingBudget = await this.budgetDbContext.Budget.AnyAsync(b => b.Name.Equals(name) && b.UserId.Equals(executionContextAccessor.GetUserId()));
+		return !anyExistingBudget;
```
Comment (pzajaczkowski) [inline src/modules/budget/application/Budget/CreatingBudget/CreateBudgetValidator.cs:53]: This negation is confusing. Method is called AnyExisitngBudget but it returns true when there aren't any matching results xD. Try to move this negation or rename method pls. Additionaly you can pass cancelation to this method and then AnyAsync.

**Review by pzajaczkowski — CHANGES_REQUESTED** [review]
I remember Mateusz G, saying something about grouping usings, to keep things more readable. That's why there were empty lines probably. Idk if it is intentional that you deleted them tho.

**Review by pzajaczkowski — APPROVED** [review]

File: src/modules/budget/application/Budget/CreatingBudget/CreateBudgetValidator.cs [hunk 4]
```diff
@@ -14,12 +16,40 @@ public class CreateBudgetValidator : AbstractValidator<CreateBudget>
...
+	public CreateBudgetValidator(BudgetDbContext budgetDbContext, IExecutionContextAccessor executionContextAccessor)
 	{
 		this.budgetDbContext = budgetDbContext;
-		this.RuleFor(budget => budget.Id).NotEmpty().NotNull();
-		this.RuleFor(budget => budget.Name).NotEmpty().NotNull().WithMessage("{PropertyName} must not be empty");
-		this.RuleFor(budget => new { budget.Period.StartDate, budget.Period.EndDate }).Must(x => x.StartDate < x.EndDate).WithMessage("The start date must be earlier than the end date");
-		this.RuleFor(budget => budget.Name).Must(x => !this.budgetDbContext.Budget.Any(b => b.Name.Equals(x))).WithMessage("{PropertyName} already exists. Choose a different name");
+
+		this.RuleFor(budget => budget.Id)
+		.NotEmpty()
```
Comment (mateuszmuszynski) [inline src/modules/budget/application/Budget/CreatingBudget/CreateBudgetValidator.cs:25]: Format code

**Review by mateuszmuszynski — CHANGES_REQUESTED** [review]

**Review by mateuszmuszynski — APPROVED** [review]

**Review by DawidBski — APPROVED** [review]

**Pull request merged**